File: src/voice/wake.ts
/**
 * Deciding whether you were talking to Loaf.
 *
 * The recogniser matches phrases; it has no idea which of them were addressed
 * to a small cat on your desktop. That decision is here, and it is here rather
 * than in Rust because it is the part with real edge cases: someone says the
 * wake word and then nothing, or says a command four minutes later, or says
 * "hey Loaf" twice. Those are cheap to test and expensive to get wrong.
 *
 * THE SHAPE. Say the wake word, get a short window, say a command. Acting on
 * a command RE-OPENS the window rather than closing it, so a follow-up needs
 * no second wake word — "open Notepad" is usually followed by something else,
 * and a wake word between every step is what makes assistants tiring.
 *
 * SILENCE IS WHAT CLOSES IT. The window is only ever extended by Loaf ACTING;
 * a spell of nothing ends it. That matters, because a window that stayed open
 * on its own would be indistinguishable, from the outside, from an assistant
 * always acting on what it hears.
 *
 * WHAT THIS DOES NOT PROTECT AGAINST, said plainly: the microphone is open the
 * whole time this is switched on. The gate decides what is ACTED ON, not what
 * is heard. The thing that makes always-on listening defensible is the closed
 * grammar in `phrases.ts` — a vocabulary that cannot transcribe a conversation
 * — not this file. See `wake.rs`.
 */

/**
 * What Loaf answers to when nobody has chosen anything else.
 *
 * "loaf" alone is included because the recogniser reliably drops the "hey",
 * and a wake word that only works when every syllable is heard is one people
 * give up on. The cost is that the name of the app, said in conversation,
 * opens the window — which is why the window alone does nothing and a command
 * still has to follow it.
 */
export const DEFAULT_WAKE_WORDS: readonly string[] = [
  "hey loaf",
  "ok loaf",
  "okay loaf",
  "loaf",
];

/** Kept for the tests and callers that predate custom wake words. */
export const WAKE_WORDS = DEFAULT_WAKE_WORDS;

/**
 * How long after the wake word a command still counts.
 *
 * Ten seconds rather than eight: the recogniser itself takes a moment to
 * settle after the wake word, and the gap between "he heard me" and "I have
 * thought of what to say" is longer than it feels when you are the one who
 * wrote the command list.
 */
export const COMMAND_WINDOW_MS = 10_000;

/** Bounds on a custom wake word, in characters. */
export const MIN_WAKE_LENGTH = 3;
export const MAX_WAKE_LENGTH = 24;

function normalise(raw: string): string {
  return raw.toLowerCase().replace(/[.,!?]+/g, " ").replace(/\s+/g, " ").trim();
}

/**
 * Clean up a wake word somebody typed, or refuse it.
 *
 * Refused rather than corrected, because the failure mode of a bad wake word is
 * silent: you say it, nothing happens, and there is no way to tell whether the
 * word was wrong or the microphone was. Better to reject it at the point it is
 * typed, where the reason can be shown.
 *
 * A single very short word is refused because a one-syllable wake word fires
 * constantly on ordinary speech, and a wake word that triggers all day is a
 * microphone with extra steps.
 */
export function normaliseWakeWord(raw: string): string | null {
  const cleaned = normalise(raw).replace(/[^a-z0-9 ]/g, "").trim();
  if (cleaned.length < MIN_WAKE_LENGTH || cleaned.length > MAX_WAKE_LENGTH) return null;
  if (!/[a-z]/.test(cleaned)) return null;
  // More than three words is a sentence, and a sentence has to be said the same
  // way twice to work.
  if (cleaned.split(" ").length > 3) return null;
  return cleaned;
}

/**
 * The words Loaf will answer to.
 *
 * A custom word REPLACES the defaults rather than joining them. Someone who
 * renamed their pet does not want it still answering to "loaf", and leaving
 * both in doubles the chances of a false wake.
 */
export function wakeWordsFor(custom?: string | null): readonly string[] {
  const chosen = custom ? normaliseWakeWord(custom) : null;
  if (chosen === null) return DEFAULT_WAKE_WORDS;
  // "hey X" as well as bare "X", for the same reason the defaults include both.
  return chosen.startsWith("hey ") ? [chosen, chosen.slice(4)] : [`hey ${chosen}`, chosen];
}

/** True when the whole phrase was just a wake word and nothing else. */
export function isWakeWord(text: string, words: readonly string[] = DEFAULT_WAKE_WORDS): boolean {
  return words.includes(normalise(text));
}

/**
 * Remove a leading wake word, if there is one.
 *
 * The grammar does not currently contain combined phrases like "hey loaf open
 * notepad" — that would double its size and a bigger grammar recognises
 * everything slightly worse. This handles them anyway, so adding them later is
 * a change to the phrase list alone.
 */
export function stripWakeWord(
  text: string,
  words: readonly string[] = DEFAULT_WAKE_WORDS,
): string {
  const t = normalise(text);
  // Longest first, or "okay loaf" leaves "loaf" behind.
  for (const wake of [...words].sort((a, b) => b.length - a.length)) {
    if (t === wake) return "";
    if (t.startsWith(`${wake} `)) return t.slice(wake.length + 1).trim();
  }
  return t;
}

/** What the caller should do about one thing that was heard. */
export interface WakeVerdict {
  /** A command to act on, or null. */
  readonly command: string | null;
  /** Whether Loaf is now waiting for a command. Drives the indicator. */
  readonly awake: boolean;
  /** True only on the utterance that woke it, so it can chirp once. */
  readonly justWoke: boolean;
}

/**
 * The gate itself.
 *
 * `now` is injected so the tests can move time without waiting for it, which
 * is the only way to test a rule whose entire content is a deadline.
 */
export class WakeGate {
  private awakeUntil = 0;

  /** Changed when the user picks a different wake word; no restart needed. */
  words: readonly string[] = DEFAULT_WAKE_WORDS;

  constructor(
    private readonly now: () => number = () => Date.now(),
    private readonly windowMs: number = COMMAND_WINDOW_MS,
    words: readonly string[] = DEFAULT_WAKE_WORDS,
  ) {
    this.words = words;
  }

  /** Whether a command said right now would be acted on. */
  get isAwake(): boolean {
    return this.now() < this.awakeUntil;
  }

  /** Milliseconds left in the window, for a countdown ring. Never negative. */
  get remainingMs(): number {
    return Math.max(0, this.awakeUntil - this.now());
  }

  heard(text: string): WakeVerdict {
    const spoken = normalise(text);
    if (spoken.length === 0) {
      return { command: null, awake: this.isAwake, justWoke: false };
    }

    // A combined "hey loaf, open notepad" acts immediately and does not leave
    // the window open behind it: the command it carried was the whole point.
    const stripped = stripWakeWord(spoken, this.words);
    const wasAddressed = stripped !== spoken;

    if (wasAddressed && stripped.length > 0) {
      // Addressed directly, so it acts — and then stays open for a follow-up
      // on the same terms as any other command.
      this.awakeUntil = this.now() + this.windowMs;
      return { command: stripped, awake: true, justWoke: false };
    }

    if (isWakeWord(spoken, this.words)) {
      const alreadyAwake = this.isAwake;
      this.awakeUntil = this.now() + this.windowMs;
      // justWoke is false on a repeat, so saying it twice does not chirp twice.
      return { command: null, awake: true, justWoke: !alreadyAwake };
    }

    if (this.isAwake) {
      // The window RE-OPENS rather than closing, so a second instruction needs
      // no second wake word. "Open Notepad" is usually followed by something
      // else, and making people say the wake word between every step is the
      // thing that makes assistants tiring to use.
      //
      // It re-opens rather than never closing: silence still ends it after
      // `windowMs`, so a conversation that has moved on does not leave Loaf
      // acting on the room.
      this.awakeUntil = this.now() + this.windowMs;
      return { command: spoken, awake: true, justWoke: false };
    }

    // Heard, and deliberately ignored. This is the common case while the
    // microphone is open and nobody is talking to the cat.
    return { command: null, awake: false, justWoke: false };
  }

  /** Close the window early — used when a command is cancelled or Loaf sleeps. */
  reset(): void {
    this.awakeUntil = 0;
  }
}

/**
 * The wake words added to the spoken vocabulary.
 *
 * Kept separate from `spokenPhrases` because they are only worth listening for
 * when always-on listening is switched on; push-to-talk already knows you are
 * talking to it, and hearing "hey Loaf" through a button press would just be a
 * phrase that does nothing.
 */
export function wakePhrases(custom?: string | null): readonly string[] {
  return [...wakeWordsFor(custom)];
}

File: src-tauri/src/wake.rs
//! Listening all the time, for the two words that mean you are talking to Loaf.
//!
//! THIS IS THE ONE FEATURE THAT CONTRADICTS THE PITCH, SO IT IS BUILT TO ARGUE
//! ITS OWN CASE. Loaf's claim is that it does not watch you. A wake word means
//! a microphone that is on whenever the app is, which is the most intrusive
//! thing this app could possibly do. Three things make it defensible, and all
//! three are load-bearing rather than decorative:
//!
//!  1. IT IS OFF UNTIL SOMEONE TURNS IT ON. There is no default, no "try it and
//!     see", and no first-run prompt that gets clicked through.
//!  2. IT STAYS ON THE MACHINE. The same `SpeechRecognitionListConstraint` as
//!     `speech.rs`, for the same reason: a continuous session with no
//!     constraints would be continuous DICTATION, streaming everything said in
//!     the room to Microsoft. This module refuses an empty phrase list exactly
//!     as `speech.rs` does, and that refusal is the whole safety property.
//!  3. IT CAN ONLY HEAR ITS OWN GRAMMAR. A closed vocabulary physically cannot
//!     transcribe a conversation. Someone talking near a listening Loaf
//!     produces "no match", not a transcript. This is the strongest privacy
//!     property here and it is a consequence of the design rather than a
//!     promise anyone has to trust.
//!
//! WHAT IS NOT DECIDED HERE. Whether a phrase followed the wake word, and
//! whether it came soon enough to count, is `src/voice/wake.ts` — pure, tested,
//! and easy to change. This file starts a session, forwards what it heard, and
//! stops. Putting the timing rules in Rust would have made the one part with
//! real edge cases the one part that cannot be tested.
//!
//! THE BUG THIS FILE USED TO HAVE, RECORDED SO IT IS NOT REINTRODUCED. WinRT's
//! `SpeechContinuousRecognitionSession` stops ITSELF after a period of
//! silence, by default. That is not an edge case for an always-on wake word —
//! it is the normal state, since most of the time nobody is talking to Loaf.
//! The first version of this file never set `AutoStopSilenceTimeout` and never
//! listened for the session's own `Completed` event, so once the OS silently
//! stopped the session, `LISTENING` stayed `true`, the closet still said
//! "Always", and nothing was actually listening — forever, with no error and
//! no way for the user to know short of trying the wake word and getting
//! nothing back. Fixed two ways: the silence timeout is set to effectively
//! never, and `Completed` is handled by restarting the same session rather
//! than leaving it dead.

use std::sync::atomic::{AtomicBool, Ordering};

/// Whether a session is running. Read by the frontend to render the indicator.
static LISTENING: AtomicBool = AtomicBool::new(false);

/// Emitted for every phrase the continuous session matches.
pub const HEARD_EVENT: &str = "loaf://voice/heard";
/// Emitted when the session stops on its own, so the UI cannot show a
/// microphone that is not actually open.
pub const STOPPED_EVENT: &str = "loaf://voice/stopped";
/// Emitted when the recogniser heard something and REJECTED it.
///
/// MEASURED, NOT GUESSED: on the machine this was written on, saying the wake
/// word produced exactly one result — empty text, confidence `Rejected` — and
/// the handler below dropped it, correctly, because acting on a rejected
/// result means acting on a sentence nobody said. The consequence was that
/// "Loaf is not listening" and "Loaf heard you and was not sure" were the same
/// silence, and no amount of trying again could tell them apart.
///
/// This carries NO text, deliberately. It exists so the UI can say "I heard
/// something" and nothing more; a rejected transcription is not evidence of
/// what was said and must never reach the command parser.
pub const REJECTED_EVENT: &str = "loaf://voice/rejected";

#[derive(Debug, Clone, serde::Serialize)]
pub struct HeardPhrase {
    pub text: String,
    pub confidence: String,
}

pub fn is_listening() -> bool {
    LISTENING.load(Ordering::SeqCst)
}

/// Refused rather than run: an empty grammar means continuous dictation, and
/// continuous dictation means a hot microphone streaming to a server.
pub const NO_PHRASES: &str = "Loaf had no phrases to listen for, so it did not start listening. \
     Listening without them would mean Windows' online recogniser.";

#[cfg(windows)]
pub use imp::{start, stop};

#[cfg(not(windows))]
pub fn start<R: tauri::Runtime>(
    _app: tauri::AppHandle<R>,
    _phrases: Vec<String>,
) -> Result<(), String> {
    Err("Always-on listening is Windows-only for now.".into())
}

#[cfg(not(windows))]
pub fn stop() {}

#[cfg(windows)]
mod imp {
    use super::{HeardPhrase, HEARD_EVENT, LISTENING, NO_PHRASES, REJECTED_EVENT, STOPPED_EVENT};
    use std::sync::atomic::Ordering;
    use std::sync::{Arc, Mutex};
    use tauri::Emitter;
    use windows::core::HSTRING;
    use windows::Foundation::Collections::IIterable;
    use windows::Foundation::TimeSpan;
    use windows::Foundation::TypedEventHandler;
    use windows::Media::SpeechRecognition::{
        SpeechContinuousRecognitionCompletedEventArgs,
        SpeechContinuousRecognitionResultGeneratedEventArgs, SpeechContinuousRecognitionSession,
        SpeechRecognitionConfidence, SpeechRecognitionListConstraint, SpeechRecognitionResult,
        SpeechRecognitionResultStatus, SpeechRecognizer,
    };
    use windows::Win32::System::Com::{CoInitializeEx, CoUninitialize, COINIT_MULTITHREADED};

    /// WinRT's own silence auto-stop, set as close to "never" as the type
    /// allows: 24 hours, in the 100-nanosecond ticks `TimeSpan` counts in.
    /// Always-on listening is meant to run for as long as the setting is on,
    /// not for as long as someone happens to keep talking.
    fn effectively_never() -> TimeSpan {
        TimeSpan {
            Duration: 24 * 60 * 60 * 10_000_000,
        }
    }

    /// Set to ask the listening thread to wind the session up.
    static STOP: std::sync::atomic::AtomicBool = std::sync::atomic::AtomicBool::new(false);

    struct Apartment(bool);

    impl Apartment {
        fn enter() -> Self {
            let hr = unsafe { CoInitializeEx(None, COINIT_MULTITHREADED) };
            Apartment(hr.is_ok())
        }
    }

    impl Drop for Apartment {
        fn drop(&mut self) {
            if self.0 {
                unsafe { CoUninitialize() };
            }
        }
    }

    fn confidence_name(c: SpeechRecognitionConfidence) -> &'static str {
        match c {
            SpeechRecognitionConfidence::High => "high",
            SpeechRecognitionConfidence::Medium => "medium",
            SpeechRecognitionConfidence::Low => "low",
            _ => "rejected",
        }
    }

    /// The best real phrase behind a rejected result, if there is one.
    ///
    /// Rejected results carry their candidates, and on this machine the wake
    /// word was reliably among them while the primary text was empty. Only
    /// phrases from the compiled grammar can appear here — see the note at the
    /// call site for why that is what makes reading them safe.
    fn best_alternate(result: &SpeechRecognitionResult) -> Option<String> {
        let alternates = result.GetAlternates(5).ok()?;
        for i in 0..alternates.Size().ok()? {
            let Ok(alt) = alternates.GetAt(i) else {
                continue;
            };
            let text = alt.Text().map(|t| t.to_string()).unwrap_or_default();
            if !text.trim().is_empty() {
                return Some(text);
            }
        }
        None
    }

    pub fn start<R: tauri::Runtime>(
        app: tauri::AppHandle<R>,
        phrases: Vec<String>,
    ) -> Result<(), String> {
        if phrases.is_empty() {
            return Err(NO_PHRASES.into());
        }
        if LISTENING.swap(true, Ordering::SeqCst) {
            // Already on. Starting twice would leave a session nothing can stop.
            return Ok(());
        }
        STOP.store(false, Ordering::SeqCst);

        // A dedicated thread, because the recogniser and its session are COM
        // objects that must live and die on one apartment. Handing them across
        // threads is how this would become an intermittent crash rather than a
        // feature.
        let (ready_tx, ready_rx) = std::sync::mpsc::channel::<Result<(), String>>();
        std::thread::spawn(move || {
            let _apartment = Apartment::enter();
            match run(&app, phrases, &ready_tx) {
                Ok(()) => {}
                Err(e) => {
                    // If it failed before reporting, the caller is still
                    // waiting; if after, the UI needs to stop showing a
                    // microphone that is not open.
                    let _ = ready_tx.send(Err(e.clone()));
                    let _ = app.emit(STOPPED_EVENT, e);
                }
            }
            LISTENING.store(false, Ordering::SeqCst);
            let _ = app.emit(STOPPED_EVENT, String::new());
        });

        match ready_rx.recv_timeout(std::time::Duration::from_secs(20)) {
            Ok(Ok(())) => Ok(()),
            Ok(Err(e)) => {
                LISTENING.store(false, Ordering::SeqCst);
                Err(e)
            }
            Err(_) => {
                LISTENING.store(false, Ordering::SeqCst);
                STOP.store(true, Ordering::SeqCst);
                Err("Speech did not start in time.".into())
            }
        }
    }

    fn run<R: tauri::Runtime>(
        app: &tauri::AppHandle<R>,
        phrases: Vec<String>,
        ready: &std::sync::mpsc::Sender<Result<(), String>>,
    ) -> Result<(), String> {
        let recognizer = SpeechRecognizer::new().map_err(|e| {
            format!(
                "Windows speech is unavailable. A speech language pack may not be installed. ({e})"
            )
        })?;

        let words: Vec<HSTRING> = phrases.iter().map(HSTRING::from).collect();
        let iterable = IIterable::<HSTRING>::try_from(words).map_err(|e| e.to_string())?;
        let constraint =
            SpeechRecognitionListConstraint::Create(&iterable).map_err(|e| e.to_string())?;
        recognizer
            .Constraints()
            .map_err(|e| e.to_string())?
            .Append(&constraint)
            .map_err(|e| e.to_string())?;

        let compiled = recognizer
            .CompileConstraintsAsync()
            .and_then(|op| op.get())
            .map_err(|e| e.to_string())?;
        let status = compiled.Status().map_err(|e| e.to_string())?;
        if status != SpeechRecognitionResultStatus::Success {
            return Err(format!("Speech could not start ({status:?})."));
        }

        let session: SpeechContinuousRecognitionSession = recognizer
            .ContinuousRecognitionSession()
            .map_err(|e| e.to_string())?;
        // See the module doc comment: without this, ordinary silence — the
        // normal state of an always-on wake word — stops the session on its
        // own.
        session
            .SetAutoStopSilenceTimeout(effectively_never())
            .map_err(|e| e.to_string())?;

        let sink = app.clone();
        session
            .ResultGenerated(&TypedEventHandler::<
                SpeechContinuousRecognitionSession,
                SpeechContinuousRecognitionResultGeneratedEventArgs,
            >::new(move |_, args| {
                let Some(args) = args.as_ref() else {
                    return Ok(());
                };
                if let Ok(result) = args.Result() {
                    let text = result.Text().map(|t| t.to_string()).unwrap_or_default();
                    let confidence = result
                        .Confidence()
                        .map(confidence_name)
                        .unwrap_or("rejected");
                    if !text.trim().is_empty() && confidence != "rejected" {
                        let _ = sink.emit(
                            HEARD_EVENT,
                            HeardPhrase {
                                text,
                                confidence: confidence.to_string(),
                            },
                        );
                    } else if let Some(alt) = best_alternate(&result) {
                        // THE BUG THIS FIXES, MEASURED ON A REAL MACHINE.
                        // Saying the wake word produced a top result of EMPTY
                        // TEXT with confidence `Rejected` — and the wake word
                        // sitting right there in the alternates:
                        //
                        //   [0] ""          raw 0.000  Rejected
                        //   [1] "loaf"      raw 0.0056 Low
                        //   [2] "hey loaf"  raw 0.0056 Low
                        //
                        // This handler only ever read `Text()`, so it read the
                        // empty string and threw the match away. Every "I said
                        // hey loaf and nothing happened" was this.
                        //
                        // WHY READING ALTERNATES IS STILL SAFE, which is the
                        // only question that matters here: the alternates come
                        // from the SAME compiled list constraint as the primary
                        // result. Every one of them is a phrase this app put in
                        // the grammar itself. There is no path by which reading
                        // them can produce a sentence that was not already in
                        // the vocabulary, so the closed-grammar property this
                        // whole module rests on is untouched.
                        //
                        // It is marked `weak` rather than passed off as a
                        // normal result, because the frontend must NOT act on
                        // one of these as a command. `voice/wake.ts` accepts a
                        // weak match only as the wake word, where being wrong
                        // costs a bubble saying "Mm?" and nothing else.
                        let _ = sink.emit(
                            HEARD_EVENT,
                            HeardPhrase {
                                text: alt,
                                confidence: "weak".to_string(),
                            },
                        );
                    } else {
                        let _ = sink.emit(REJECTED_EVENT, ());
                    }
                }
                Ok(())
            }))
            .map_err(|e| e.to_string())?;

        // `Completed` fires whenever the session stops running — deliberately
        // (StopAsync, below) or not (silence, an audio device change, another
        // app taking the microphone). Recording that it fired, and why, is
        // what turns "silently dead forever" into "restarted" or, after
        // repeated failures, an actual reported reason instead of a mystery.
        let died = Arc::new(std::sync::atomic::AtomicBool::new(false));
        let last_status: Arc<Mutex<Option<SpeechRecognitionResultStatus>>> =
            Arc::new(Mutex::new(None));
        let died_for_handler = Arc::clone(&died);
        let status_for_handler = Arc::clone(&last_status);
        session
            .Completed(&TypedEventHandler::<
                SpeechContinuousRecognitionSession,
                SpeechContinuousRecognitionCompletedEventArgs,
            >::new(move |_, args| {
                if let Some(args) = args.as_ref() {
                    if let Ok(status) = args.Status() {
                        *status_for_handler.lock().unwrap() = Some(status);
                    }
                }
                died_for_handler.store(true, Ordering::SeqCst);
                Ok(())
            }))
            .map_err(|e| e.to_string())?;

        session
            .StartAsync()
            .and_then(|op| op.get())
            .map_err(|e| e.to_string())?;

        let _ = ready.send(Ok(()));

        // A restart budget rather than an unconditional retry loop: five
        // restarts within a rolling minute is almost certainly a transient
        // hiccup (another app briefly took the microphone); more than that is
        // more likely something actually wrong — no microphone, a driver
        // problem — and restarting forever would hide that behind a mode
        // that LOOKS on but never hears anything, the exact failure this
        // rewrite exists to fix.
        let mut restarts_in_window = 0u32;
        let mut window_started = std::time::Instant::now();

        while !STOP.load(Ordering::SeqCst) {
            if died.swap(false, Ordering::SeqCst) {
                if window_started.elapsed() > std::time::Duration::from_secs(60) {
                    restarts_in_window = 0;
                    window_started = std::time::Instant::now();
                }
                restarts_in_window += 1;
                if restarts_in_window > 5 {
                    let status = last_status.lock().ok().and_then(|mut s| s.take());
                    return Err(format!(
                        "Speech kept stopping on its own ({status:?} last time) even after \
                         several restarts, so Loaf gave up rather than loop forever."
                    ));
                }
                std::thread::sleep(std::time::Duration::from_millis(400));
                // The SAME session object, not a new one: `ResultGenerated` and
                // `Completed` are attached to the session for its whole
                // lifetime, so calling `StartAsync` again resumes it without
                // re-registering either handler.
                if let Err(e) = session.StartAsync().and_then(|op| op.get()) {
                    return Err(format!("Could not restart listening: {e}"));
                }
            }
            std::thread::sleep(std::time::Duration::from_millis(150));
        }

        // Best effort. A session that has already ended returns an error here
        // and that is not worth reporting to anyone.
        let _ = session.StopAsync().map(|op| op.get());
        Ok(())
    }

    pub fn stop() {
        STOP.store(true, Ordering::SeqCst);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The rule that can be checked anywhere: nothing starts without a
    /// vocabulary, because the alternative is continuous dictation.
    #[test]
    fn the_refusal_says_why() {
        assert!(NO_PHRASES.contains("online"));
    }

    #[test]
    fn starts_switched_off() {
        assert!(!is_listening());
    }
}
